Training24/intlist.c

```diff
@@ -0,0 +1,200 @@
+      newNode->next = list->head;
+      list->head = newNode;
+   }
+   else {
+      Node* current = list->head;
+      for (int i = 0; i < index - 1; i++) {  // Traverse to the node before the insertion point
+         current = current->next;
+      }
+      newNode->next = current->next;
+      current->next = newNode;
+   }
+   mlastError = SUCCESS;
+}
+
+// Remove the element at a specified index (zero-based)
+void RemoveAt (IntList* list, int index) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   if (index < 0 || index >= Count (list)) {
+      mlastError = INVALID_INDEX;
+      return;
+   }
+   Node* current = list->head;
+   Node* prev = NULL;
+   if (index == 0) { //If removing the head node
+      list->head = current->next;
+      free (current);
+   }
+   else {   // Traverse to the node before the one to be removed
+      for (int i = 0; i < index - 1; i++) {
+         prev = current;
+         current = current->next;
+      }
+      Node* temp = current->next;
+      if (temp != NULL) {
+         current->next = temp->next;   // Node to be removed
+         free (temp);
+      }
+      else {
+         mlastError = INVALID_INDEX;
+         return;
+      }
+   }
+   mlastError = SUCCESS;
+}
+
+// Remove the first occurrence of a specific element
+void Remove (IntList* list, int data) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   Node* current = list->head;
+   Node* prev = NULL;
+   while (current != NULL) {  // Traverse the list to find the first occurrence of the element
+      if (current->data == data) {
+         if (prev == NULL) {  // If the element to remove is found at head of the list
+            list->head = current->next;
+         }
+         else {
+            prev->next = current->next;   // The element is somewhere after the head
+         }
+         free (current);
+         mlastError = SUCCESS;
+         return;
+      }
+      prev = current;
+      current = current->next;   // Move to the next node
+   }
+   mlastError = ELEMENT_NOT_FOUND;
+}
+
+// Count the number of elements in the list
+int Count (IntList* list) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return 0;
+   }
+   int count = 0;
+   Node* current = list->head;
+   while (current != NULL) {  // Traverse the list and count nodes
+      count++;
+      current = current->next;
+   }
+   mlastError = SUCCESS;
+   return count;
+}
+
+// Get the element at a specified index
+int Get (IntList* list, int index, int* element) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return INVALID_INDEX;
+   }
+   if (index < 0 || index >= Count (list)) {
+      mlastError = INVALID_INDEX;
+      return INVALID_INDEX;
+   }
+   Node* current = list->head;
+   for (int i = 0; i < index; i++) {   // Traverse to the node at the specified index
+      current = current->next;
+   }
+   *element = current->data;
+   mlastError = SUCCESS;

```

Comment:
This is a nice way to differentiate b/w output and error. You could also return index of input if found. In this case negative values would be error codes.

Reply:
Yes this is a nice way too. But that would mean another call to get the value at the specified index.